### ISSUE TITLE
Recursive SQL Query Leads to Infinite Loop

### ISSUE THREAD

**Issue by suyZhong** [duckdb/duckdb#8667]
### What happens?

Consider the test case below:
```sql
WITH RECURSIVE x(n) AS (
  SELECT 1 
  UNION ALL 
  SELECT n+1 
  FROM x
  WHERE n IN (SELECT * FROM x)
)
SELECT * FROM x;
```
DuckDB would hang for this query, whereas PostgreSQL, MySQL, and SQLite would throw an error.

I think it is a test case exercising underspecified behavior, but is it something that could be optimized? 

### To Reproduce

```sql
WITH RECURSIVE x(n) AS (
  SELECT 1 
  UNION ALL 
  SELECT n+1 
  FROM x
  WHERE n IN (SELECT * FROM x)
)
SELECT * FROM x;
```

### OS:

ubuntu 22.04

### DuckDB Version:

v0.8.2-dev3055 b105af71bb

### DuckDB Client:

CLI

### Full Name:

Suyang Zhong

### Affiliation:

NUS

### Have you tried this on the latest `main` branch?

I have tested with a main build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by Reminiscent** [comment]
Hi, Can I try to fix this issue?

**Comment by Reminiscent** [comment]
The reason is because in the recursive CTE part. We just treat the CTE table(`x`) as a base table. So it can be referenced everywhere. But in the postgres or mysql, there are a lot of strict constraints for the CTE table. I'll try to figure out how to add some constraints for the CTE table when we handle the recursive part.

**Comment by Reminiscent** [comment]
After reviewing code snippets related to recursive CTEs, I've identified some key points regarding the issue:

In MySQL (resulting in error `ERROR 3577 (HY000): In recursive query block of Recursive Common Table Expression 'x', the recursive table must be referenced only once, and not in any subquery`) and Postgres (resulting in error `ERROR:  recursive reference to query "x" must not appear within a subquery`), there are limitations on referencing the recursive CTE table within subqueries. These databases restrict recursive CTE tables from being referenced more than once, and not within subqueries. Due to these constraints, the provided query would result in an error. However, in DuckDB, such limitations on recursive CTEs do not exist, and hence, the query would not throw a similar error.

Another potential problem arises from DuckDB's allowance of recursive CTEs lacking termination conditions, which can lead to an infinite loop. This behavior is further discussed in the context of [this comment](https://github.com/duckdb/duckdb/pull/404#issuecomment-580746437). An example query that demonstrates this is with recursive t as (select 42 as x union all select * from t) select * from t;, which also leads to an infinite loop. It might be prudent to provide more comprehensive guidance and information to users regarding this behavior, or alternatively, consider implementing additional constraints for recursive CTEs.

Based on the [PR](https://github.com/duckdb/duckdb/pull/404), so do you have any ideas about this issue? @kryonix @Mytherin

**Comment by kryonix** [comment]
Thank you for having a look @Reminiscent. I think that the infinite loop is in fact expected behavior in this case because the supposed termination check `WHERE n IN (SELECT * FROM x)` only ever sees the rows from the previous iteration of the recursive CTE. This is due to the semi-naive evaluation strategy used to evaluate recursive CTEs.

> I'll try to figure out how to add some constraints for the CTE table when we handle the recursive part.

Of course it would be possible to restrict recursive CTEs in a similar manner PostgreSQL and other DBMSs do. **BUT**, I am not a fan of this at all. I don't think that such restrictions are helpful to prevent infinite loops, for example, but they are definitively hindering in some cases (resulting in awkward workarounds).

> Another potential problem arises from DuckDB's allowance of recursive CTEs lacking termination conditions, which can lead to an infinite loop

That's true, but implementing a termination check _for recursive CTEs_ is impossible, as these CTEs are Turing-complete.

TL;DR — I don't think that this issue requires a fix.

**Comment by Teggy** [comment]
Hi all,

there is, indeed, no issue to fix here.   

The predicate `n IN (SELECT * FROM x)` is equivalent to `true` (`n` has just been placed in table `x` in the prior iteration) such that the recursive part of the recursive CTE will produce a non-empty single-row result in _each iteration_, an indication that (at least) one more iteration is needed.

If you deliberately restrict the overall size of the result, e.g., via

~~~
WITH RECURSIVE x(n) AS (
  SELECT 1 
  UNION ALL 
  SELECT n+1 
  FROM x
  WHERE n IN (SELECT * FROM x)
)
SELECT * FROM x
LIMIT 10;                   -- added
~~~

you will see the expected prefix of the (conceptually infinite) result of the CTE:

~~~
┌───────┐
│   n   │
│ int32 │
├───────┤
│     1 │
│     2 │
│     3 │
│     4 │
│     5 │
│     6 │
│     7 │
│     8 │
│     9 │
│    10 │
└───────┘
~~~

**Comment by Reminiscent** [comment]
@Teggy @kryonix I appreciate your thorough explanation. I'm in complete agreement with you regarding the design principles of recursive CTEs in DuckDB.
